Libraries/CO.CDP.AspNetCore/ErrorHandlerMiddleware.cs

```diff
@@ -0,0 +1,66 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+using System.ComponentModel.DataAnnotations;
+using System.Text.Json;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace CO.CDP.AspNetCore;
+public class ErrorHandlerMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly ILogger<ErrorHandlerMiddleware> _logger;
+    private readonly IWebHostEnvironment _environment;
+
+    public ErrorHandlerMiddleware(RequestDelegate next, ILogger<ErrorHandlerMiddleware> logger, IWebHostEnvironment environment)
+    {
+        _next = next;
+        _logger = logger;
+        _environment = environment;
+    }
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        try
+        {
+            await _next(context);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "An unexpected error occurred right");
+            await HandleExceptionAsync(context, ex);
+        }
+    }
+
+    private Task HandleExceptionAsync(HttpContext context, Exception exception)
+    {
+        var response = context.Response;
+        response.ContentType = "application/json";
+
+        var statusCode = exception switch
+        {
+            KeyNotFoundException => (int)HttpStatusCode.NotFound,
+            ArgumentException => (int)HttpStatusCode.BadRequest,
+            JsonException => StatusCodes.Status422UnprocessableEntity,
+            BadHttpRequestException => StatusCodes.Status422UnprocessableEntity,
+            _ => (int)HttpStatusCode.InternalServerError,
+        };
+
+        var problemDetails = new ProblemDetails
+        {
+            Status = statusCode,
+            Title = statusCode == (int)HttpStatusCode.InternalServerError ? "An unexpected error occurred" : exception.Message,
+            Detail = _environment.IsDevelopment() ? exception.ToString() : (statusCode == (int)HttpStatusCode.InternalServerError ? "Internal Server Error" : exception.Message),
+        };
```

Comment:
One disadvantage of doing the mapping manually like this instead of using the [built-in mechanisms](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-8.0) (`services.AddProblemDetails()` and `app.UseExceptionHandler();`) is that we lose the type:

```
{
  "type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
  "title": "Bad Request",
  "status": 400
}
```

Reply:
On the other hand, we need to pass the type of error back to the caller, as the frontend will need to map to its own error message at least for some errors. Should we be using custom `type` to pass this information back?